gemonlinedb/src/common/SystemTopology.cc

```diff
@@ -0,0 +1,99 @@
+#include "gem/onlinedb/SystemTopology.h"
+
+#include <stdexcept>
+
+#include "gem/onlinedb/detail/XMLUtils.h"
+
+XERCES_CPP_NAMESPACE_USE
+
+namespace gem {
+    namespace onlinedb {
+        namespace /* anonymous */ {
+            /**
+             * @brief Retrieves a part type from the attributes of a DOM element.
+             */
+            template<typename PartType>
+            void populatePart(PartType &part, const DOMElement *el);
+
+            template<>
+            void populatePart<PartReferenceBarcode>(PartReferenceBarcode &part,
+                                                    const DOMElement *el)
+            {
+                using namespace detail::literals;
+                part.barcode = detail::transcode(el->getAttribute("barcode"_xml));
+            }
+
+            template<>
+            void populatePart<PartReferenceSN>(PartReferenceSN &part,
+                                               const DOMElement *el)
+            {
+                using namespace detail::literals;
+                part.serialNumber =
+                    detail::transcode(el->getAttribute("serial-number"_xml));
+            }
+        } // anonymous namespace
+
+        void SystemTopology::populate(const DOMDocumentPtr &document)
+        {
+            auto topologyNode = detail::xsdGet(document, "//gem:system/gem:topology");
+            if (topologyNode == nullptr) {
+                throw std::logic_error("Invalid XML topology definition");
+            }
+
+            auto topologyEl = dynamic_cast<DOMElement *>(topologyNode);
+            auto amc13El = topologyEl->getFirstElementChild();
+
+            m_root.populate(amc13El);
+        }
+
+        void SystemTopology::AMC13Node::populate(const DOMElement *el)
+        {
+            populatePart(reference, el);
+
+            amc.clear();
+            auto amcEl = el->getFirstElementChild();
+            do {
```

Comment:
here you suppose that the AMC13 node contains at least one AMC child. This has to be checked beforehand and exception thrown if such child node is not found.

Reply:
Good catch